Handle jobs execution order properly

firecow opened the pull request.

firecow reviewed (commented):
Just review'ed everything here on github. LGTM

moberghammer commented on src/commander.ts:
```diff
@@ -3,126 +3,56 @@ import {Job} from "./job";
+            jobArgs.forEach(jobName => {
+                const job = Utils.getJobByName(jobs, jobName);
+                jobPoolMap.set(jobName, job);
+                potentialStarters = potentialStarters.concat(JobExecutor.getPastToWaitFor(jobs, stages, job));
```
Move if needs to this line only to depublicate

firecow replied:
Job pool map is also special

moberghammer reviewed (approved):
Huge PR but looks okay as far as i can tell.

bcouetil reviewed (approved):
Tested  #110, it's working.

Too big for me to understand the code. Maybe you could explain the change in philosophy, for us to detect flaws in principle ?

firecow commented:
@bcouetil Well, the pipeline and job execution was completely hacked together, and has been like that for a long time.
I've simplified job execution in the JobExecutor class, and added a lot of missing testes, to fix the issues reported.
So its not really a change in philosophy, as much as it is a move towards the current philosophy, with lots and lots of integration testing :smile:

bcouetil commented:
OK, merge when you want :wink:
